### PR TITLE
DROOLS-2655: [DMN Designer]: Decision Table: Automatically create input columns for each InputData element

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/DROOLS-2655

### CODE REVIEW

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
```
Build timed out (after 90 minutes). Marking the build as aborted.
```

Perhaps increasing the build timeout would a good idea.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by jomarko** [comment]
jenkins please retest this

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableEditorDefinitionEnricher.java [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    @Inject
+    public DecisionTableEditorDefinitionEnricher(final SessionManager sessionManager,
+                                                 final DMNGraphUtils dmnGraphUtils) {
+        this.sessionManager = sessionManager;
+        this.dmnGraphUtils = dmnGraphUtils;
+    }
+
+    public void enrichModelClass(final Optional<String> nodeUUID,
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableEditorDefinitionEnricher.java]: Please add `@Override` annotation

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableEditorDefinitionEnricher.java]: @jomarko The method does not override anything.

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableEditorDefinitionEnricher.java]: @jomarko I've moved the method out to an interface and refactored accordingly.

**Comment by manstis** [comment]
@jomarko New commit sorts alphabetically and includes the interface changes mention [earlier](https://github.com/kiegroup/kie-wb-common/pull/1955#discussion_r201308042).

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by karreiro** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by manstis** [comment]
Jenkins please retest this.

**Comment by jomarko** [comment]
@manstis I am not able to finish manual check due to [DROOLS-2721](https://issues.jboss.org/browse/DROOLS-2721)

**Comment by manstis** [comment]
@jomarko [DROOLS-2721](https://issues.jboss.org/browse/DROOLS-2721) has been merged.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/graph/DMNGraphUtilsTest.java [hunk 2]
```diff
@@ -67,15 +67,14 @@
 
     private DMNGraphUtils utils;
 
-    private ProjectDiagramImpl diagram;
-
     private GraphImpl graph;
 
     @Before
+    @SuppressWarnings("unchecked")
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/test/java/org/kie/workbench/common/dmn/client/graph/DMNGraphUtilsTest.java]: We could remove this line by changing this:
```java
// [...]
private GraphImpl graph;
// [...]
this.graph = new GraphImpl(UUID.uuid(), new GraphNodeStoreImpl());
// [...]
```
..to this:
```java
// [...]
private GraphImpl<DefinitionSet> graph;
// [...]
this.graph = new GraphImpl<>(UUID.uuid(), new GraphNodeStoreImpl());
// [...]
```

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/supplementary/java/JavaFunctionEditorDefinition.java [hunk 3]
```diff
@@ -106,22 +106,30 @@ public String getName() {
...
+            context.getContextEntry().add(classEntry);
+
+            final ContextEntry methodSignatureEntry = new ContextEntry();
+            final InformationItem methodSignatureEntryVariable = new InformationItem();
+            methodSignatureEntryVariable.setName(new Name(VARIABLE_METHOD_SIGNATURE));
+            methodSignatureEntry.setVariable(methodSignatureEntryVariable);
+            methodSignatureEntry.setExpression(new LiteralExpression());
+            context.getContextEntry().add(methodSignatureEntry);
+        });
+    }
```
Comment (karreiro) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/supplementary/java/JavaFunctionEditorDefinition.java]: There's a similar method in `PMMLFunctionEditorDefinition`.. maybe we could extract this.

Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/supplementary/java/JavaFunctionEditorDefinition.java]: @karreiro Unknowingly your proposal has unearthed a bug in this PR (if you selected "Function" as an expression type in the "grid" editor of a decision and then tried to change it to a JAVA or PMML function the resulting grid was broken). Congratulations! Refactored (and fixed) in a new commit.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableEditorDefinition.java [hunk 4]
```diff
@@ -101,45 +96,16 @@ public String getName() {
...
-
-        //Setup parent relationships
-        ic.setParent(dtable);
-        oc.setParent(dtable);
-        dr.setParent(dtable);
-        le.setParent(ic);
-        drut.setParent(dr);
-        drle.setParent(dr);
 
         return Optional.of(dtable);
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableEditorDefinition.java]: We could shorten now to `return Optional.of(new DecisionTable());`

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/FunctionEditorDefinition.java [hunk 5]
```diff
@@ -102,12 +102,22 @@ public String getName() {
     @Override
     public Optional<FunctionDefinition> getModelClass() {
         final FunctionDefinition function = new FunctionDefinition();
-        KindUtilities.setKind(function,
-                              FunctionDefinition.Kind.FEEL);
-        function.setExpression(new LiteralExpression());
+
         return Optional.of(function);
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/FunctionEditorDefinition.java]: We could shorten now to `return Optional.of(new FunctionDefinition());`

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/supplementary/BaseSupplementaryFunctionEditorDefinition.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+              listSelector,
+              translationService);
+        this.expressionEditorDefinitionsSupplier = expressionEditorDefinitionsSupplier;
+    }
+
+    @Override
+    public Optional<Context> getModelClass() {
+        final Context context = new Context();
+
+        return Optional.of(context);
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/function/supplementary/BaseSupplementaryFunctionEditorDefinition.java]: We could shorten to one line `return Optional.of(new Context());`

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/invocation/InvocationEditorDefinition.java [hunk 7]
```diff
@@ -97,23 +97,30 @@ public String getName() {
...
-        final Binding binding = new Binding();
-        binding.setParameter(parameter);
-        invocation.getBinding().add(binding);
-
-        //Setup parent relationships
-        literalExpression.setParent(invocation);
-        binding.setParent(invocation);
-        parameter.setParent(binding);
 
         return Optional.of(invocation);
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/invocation/InvocationEditorDefinition.java]: `return Optional.of(new Invocation());`

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/relation/RelationEditorDefinition.java [hunk 8]
```diff
@@ -90,22 +90,29 @@ public String getName() {
...
-        row.getExpression().add(literalExpression);
-        relation.getColumn().add(column);
-        relation.getRow().add(row);
-
-        //Setup parent relationships
-        column.setParent(relation);
-        row.setParent(relation);
-        literalExpression.setParent(row);
 
         return Optional.of(relation);
```
Comment (jomarko) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/relation/RelationEditorDefinition.java]: `return Optional.of(new Relation());`

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableEditorDefinition.java [hunk 9]
```diff
@@ -58,6 +50,7 @@
 public class DecisionTableEditorDefinition extends BaseEditorDefinition<DecisionTable, DecisionTableGridData> {
 
     private HitPolicyEditorView.Presenter hitPolicyEditor;
+    private DecisionTableEditorDefinitionEnricher enricher;
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/expressions/types/dtable/DecisionTableEditorDefinition.java]: LOL, OK so let me explain my thoughts. Originally I had ```DecisionTableEditorDefinition``` contain all the logic for the "enrichment" however I thought:
> Crikey that's a lot of logic, I know I'll move it to a dedicated class that performs the enrichment and _delegate_ to that from the definition class. It'll make Unit Testing easier too. I'll only do it for ```DecisionTableEditorDefinition``` as that's the only one that has a lot of specialised enrichment logic.

To be honest I don't see this as any different to many other places where we delegate implementation to other classes.

**Comment by manstis** [comment]
@jomarko Requested changes made in new commit. Thank-you for your review.

**Comment by manstis** [comment]
Jenkins please retest this.

**Pull request merged**